### PR TITLE
Bugfix/unr 4916 fix snapshot reloading

### PR DESCRIPTION
**Pull request by m-samiec**
#### Description
Fix snapshot reloading in v15.

#### Tests
I tested this in the editor by disabling "Delete dynamic entities" and switching auto-stop local deployment to "Never".
Then ran through numerous gyms starting and stopping, and validating that existing state persisted, startup actors were relinked correctly, not double created etc.

#### Primary reviewers
@yumnuska @alastairdglennie @samiwh

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -2665,7 +2665,6 @@ void USpatialNetDriver::TryFinishStartup()
 		}
 		else if (VirtualWorkerTranslator.IsValid() && !VirtualWorkerTranslator->IsReady())
 		{
-			GlobalStateManager->QueryTranslation();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2668]: No longer required as vw translation is part of well known gdk components, which all server workers have interest in. Additionally this query was invalidating the local server's state that had authority over the translation and was manually shorting updates.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp [hunk 2]
```diff
@@ -334,6 +334,8 @@ TArray<FWorkerComponentData> EntityFactory::CreatePartitionEntityComponents(cons
 	Components.Add(Metadata(FString::Format(TEXT("PartitionEntity:{0}"), { VirtualWorker })).CreateMetadataData());
 	Components.Add(InterestFactory->CreatePartitionInterest(LbStrategy, VirtualWorker, bDebugContextValid).CreateInterestData());
 	Components.Add(AuthorityDelegation(DelegationMap).CreateAuthorityDelegationData());
+	Components.Add(Persistence().CreatePersistenceData());
+	Components.Add(ComponentFactory::CreateEmptyComponentData(SpatialConstants::PARTITION_SHADOW_COMPONENT_ID));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:338]: @yumnuska @samiwh thoughts on adding the partition shadow component to the well known snapshot partition also? I didn't for now because it didn't seem necessary, but not hugely opinionated,

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp [hunk 3]
```diff
@@ -98,13 +98,6 @@ bool SpatialVirtualWorkerTranslator::IsValidMapping(Schema_Object* Object) const
 // a worker first becomes authoritative for the mapping.
 void SpatialVirtualWorkerTranslator::ApplyMappingFromSchema(Schema_Object* Object)
 {
-	if (!IsValidMapping(Object))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp:101]: I don't believe this check is necessary anymore - was originally added https://github.com/spatialos/UnrealGDK/commit/f07850cd89746a4cc04063e138894ac0804d10b1

**Comment by yumnuska** [comment]
I haven't pulled to test yet (hopefully today, just leaving a note in case I don't get to it) but this makes sense to me based on what I was expecting to need to do.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp (hunk 3 above)
Comment (yumnuska) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp:101]: I think it would still be worth validating that the mapping matches the number of workers expected for the deployment, although I see that you've got that covered elsewhere too.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp (hunk 2 above)
Comment (yumnuska) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:338]: I don't think it's strictly necessary, but IMO its less confusing to include it (otherwise it might seem like it was omitted by mistake?). I can also imagine it coming in handy at some point.

Comment (yumnuska) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:338]: one way it might come in handy is during a snapshot migration -- it means that we can assume that all partitions referred to by the AuthorityDelegation components will exist in the snapshot without having to special case EntityId == 4.

Comment (yumnuska) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:338]: all server partitions that is, which means we can assume other partitions are "client" partitions

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp [hunk 4]
```diff
@@ -54,8 +54,24 @@ void SpatialVirtualWorkerTranslationManager::AuthorityChanged(const Worker_Compo
 
 	UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log, TEXT("This worker now has authority over the VirtualWorker translation."));
 
-	// We need to create partition entities before we start assigning virtual workers.
-	SpawnPartitionEntitiesForVirtualWorkerIds();
+	int32 ExistingTranslatorMappingCount = Translator->GetMappingCount();
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp:57]:
```suggestion
	const int32 ExistingTranslatorMappingCount = Translator->GetMappingCount();
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp [hunk 5]
```diff
@@ -71,6 +87,18 @@ void SpatialVirtualWorkerTranslationManager::SpawnPartitionEntitiesForVirtualWor
 	}
 }
 
+void SpatialVirtualWorkerTranslationManager::ReclaimPartitionEntities()
+{
+	Partitions.Empty();
+	for (VirtualWorkerId VirtualWorkerId : VirtualWorkersToAssign)
+	{
+		Worker_PartitionId PartitionId = Translator->GetPartitionEntityForVirtualWorker(VirtualWorkerId);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp:95]:
```suggestion
		const Worker_PartitionId PartitionId = Translator->GetPartitionEntityForVirtualWorker(VirtualWorkerId);
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp [hunk 6]
```diff
@@ -71,6 +87,18 @@ void SpatialVirtualWorkerTranslationManager::SpawnPartitionEntitiesForVirtualWor
 	}
 }
 
+void SpatialVirtualWorkerTranslationManager::ReclaimPartitionEntities()
+{
+	Partitions.Empty();
+	for (VirtualWorkerId VirtualWorkerId : VirtualWorkersToAssign)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp:93]:
```suggestion
	for (const VirtualWorkerId VirtualWorkerId : VirtualWorkersToAssign)
```

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp:93]: maybe it doesn't like const in the loop def

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp [hunk 7]
```diff
@@ -71,6 +87,18 @@ void SpatialVirtualWorkerTranslationManager::SpawnPartitionEntitiesForVirtualWor
...
+	{
+		Worker_PartitionId PartitionId = Translator->GetPartitionEntityForVirtualWorker(VirtualWorkerId);
+		check(PartitionId != SpatialConstants::INVALID_ENTITY_ID);
+		Partitions.Emplace(PartitionInfo{ PartitionId, VirtualWorkerId, SpatialConstants::INVALID_ENTITY_ID });
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp:97]: Have u run clang-format? Might've expected it to expand these braces a bit

**Review by alastairdglennie — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp (hunk 2 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:338]: Cool - have added to the snapshot partition.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp (hunk 7 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp:99]: Yep, just the way it likes it.

File: SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp [hunk 8]
```diff
@@ -210,6 +210,7 @@ void FLocalDeploymentManager::TryStartLocalDeployment(FString LaunchConfig, FStr
 													  const LocalDeploymentCallback& CallBack)
 {
 	RuntimeStartTime = FDateTime::Now();
+	bRedeployRequired = false;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:213]: Driveby where the "Redeploy" flag was never set to false after being enabled. This resulted in deployments restarting even when not desired. @joshuahuburn

File: SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp [hunk 9]
```diff
@@ -976,7 +975,7 @@ void FSpatialGDKEditorToolbarModule::LaunchInspectorWebpageButtonClicked()
 		});
 
 		InspectorProcess->OnCanceled().BindLambda([this] {
-			if (InspectorProcess->GetReturnCode() != SpatialGDKServicesConstants::ExitCodeSuccess)
+			if (InspectorProcess.IsSet() && InspectorProcess->GetReturnCode() != SpatialGDKServicesConstants::ExitCodeSuccess)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp:978]: Driveby to fix a crash on shutdown

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp (hunk 3 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp:101]: Yep, additionally the ValidMapping was only returning true if a of the physical worker names in the data matches that of the local worker, which isn't valid for snapshot reloading.

**Review by joshuahuburn — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp [hunk 10]
```diff
@@ -141,6 +141,7 @@ bool CreateGlobalStateManager(Worker_SnapshotOutputStream* OutputStream)
 	Components.Add(Position(DeploymentOrigin).CreatePositionData());
 	Components.Add(Metadata(TEXT("GlobalStateManager")).CreateMetadataData());
 	Components.Add(Persistence().CreatePersistenceData());
+	Components.Add(ComponentFactory::CreateEmptyComponentData(SpatialConstants::PARTITION_SHADOW_COMPONENT_ID));
```
Comment (yumnuska) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp:144]: Why are you putting this on the GSM instead of the snapshot partition entity?

**Comment by yumnuska** [comment]
I've verified on NWX that this does what I expect -- the server partitions are saved to the snapshot (verified by inspection), and they are correctly reclaimed when loading from a snapshot.

**Comment by yumnuska** [comment]
I'm noticing something else worth tracking outside this change tho. The AuthroityDelegation can contain entries for client partitions, and at some point those need to be cleaned up. I'm not sure where the best place for this to happen is, probably worth a  catchup later.

**Comment by yumnuska** [comment]
image
for posterity. note the delgations for partition 10, which was the entity ID of the player controller when the snapshot was taken. (Arguably, most of those entities shouldn't even be in the snapshot tho)

**Comment by yumnuska** [comment]
image
one more capture, showing the partition shadow on the GSM instead of the snapshot partition entity like I expect.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp (hunk 10 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SnapshotGenerator/SpatialGDKEditorSnapshotGenerator.cpp]: ugh - fixed

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp (hunk 3 above)
Comment (yumnuska) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslator.cpp:101]: IsValidMapping can be removed now, right?

**Review by yumnuska — APPROVED** [review]
